core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -642,7 +642,14 @@ class ReplicaManager(val config: KafkaConfig,
           (entriesPerPartition, Map.empty)
         else
           entriesPerPartition.partition { case (topicPartition, records) =>
-            getPartitionOrException(topicPartition).hasOngoingTransaction(records.firstBatch().producerId())
+            // Produce requests (only requests that require verification) should only have one batch in "batches" but check all just to be safe.
+            val transactionalBatches = records.batches.asScala.filter(batch => batch.hasProducerId && batch.isTransactional)
```

Comment:
Is it valid for `isTransactional` to be `true` but `hasProducerId` to be `false`?

Reply:
it would be the other way around, we can have a producer id but not be transactional with idempotent producers.